### DISCUSSION TITLE
seamless-blend composite ignores transparency

### DISCUSSION THREAD

**Discussion by druzhynin-oleksii** [ImageMagick/ImageMagick#6578] [Help] [answered]
### ImageMagick version

7.1.1-15

### Operating system, version and so on

Linux

### Description

I am trying to fit image with transparent pixels into another:

`magick -verbose tararget_iget_img.field512.png obj_square.png  obj_square.mask.png  -define compose:args=30000x0.0000000000000000000003+200 -compose seamless-blend  -geometry +30+30 -composite obj_in_tile.png`

obj_square.mask.png 
![image](https://github.com/ImageMagick/ImageMagick/assets/71090115/847139d9-57d8-457c-a8ef-cb6a31e8b80d)
obj_square.png 
![image](https://github.com/ImageMagick/ImageMagick/assets/71090115/cb6316fd-d20d-45ad-a8f7-05197249ac70)
tararget_iget_img.field512.png 
![image](https://github.com/ImageMagick/ImageMagick/assets/71090115/f7f799de-0207-4ae7-91e7-360e192b9136)

obj_in_tile.png
![image](https://github.com/ImageMagick/ImageMagick/assets/71090115/444cc4e9-c4d5-46d4-9c71-0ed6a79d4db9)

**Comment by snibgo** [comment]
Please show your three input images.

Your title refers to `seamless-blend` but your code uses `saliency-blend`. These are related, but not the same

As far as I know, neither operation is defined for fully-transparent pixels. In theory they could be defined for partially-transparent pixels, but I don't think they are actually defined, because the "most salient" of images with transparency is not well-defined.

You might try copying the alpha channel of each image into a meta channel. Then do the composite operation (which replaces the 2 or 3 input images with one output image). Finally, copy the meta channel into the alpha channel.

**Comment by druzhynin-oleksii** [comment]
Thanks a lot for the answer. I've updated the topic starting message.

But overlay works fine: `magick -verbose target_img.field512.png \( obj_square.png -background Black -alpha Background  \)  -compose overlay -geometry +30+30   -composite obj_in_tile.png`

![image](https://github.com/ImageMagick/ImageMagick/assets/71090115/9b2aa496-c631-4c3b-92b3-735fda0710f9)

**Comment by snibgo** [comment]
Thanks for the images. You don't say which one is which. I suppose they are obj_square.mask.png, obj_square.png and target_img.field512.png, in that order.

What result do you want? If you want the truck against the background of `target_img.field512.png`, then I suggest:

```
magick target_img.field512.png obj_square.png -compose Over -composite x.png
```
`-compose Over` is the default, so you don't need to explicitly include this.

What is wrong with that result? What difference do you want from `-compose seamless-blend`?

**Comment by druzhynin-oleksii** [comment]
Hi Alan

I want to have a style adjustment as in https://github.com/owenzlz/DeepImageBlending for dataset synthesis: 
`python run.py        --source_file data/obj_square.png        --mask_file data/obj_square.mask.png        --target_file data/tardet_img01.tile256.png        --output_dir results/2        --ss 400 --ts 512 --x 256 --y 256        --gpu 0  --num_steps 1000 --save_video True`

![image](https://github.com/ImageMagick/ImageMagick/assets/71090115/ed9b620f-05be-4bc8-94a6-d07973b9deec)
![image](https://github.com/ImageMagick/ImageMagick/assets/71090115/cf7528d1-3970-4546-95e1-5fb41db0851d)

The example in https://github.com/ImageMagick/ImageMagick/discussions/4712 looks very close to this.

**Comment by druzhynin-oleksii** [comment]
I am wondering why implicit adding of background doesn't work

obj_square.back.png
![image](https://github.com/ImageMagick/ImageMagick/assets/71090115/11e52539-afb3-4e69-946c-aa801bbed834)

`magick -verbose target_img.field512.png obj_square.back.png  obj_square.mask.png  -define compose:args=30000x0.0000000000000000000003+200 -compose seamless-blend  -geometry +30+30 -composite obj_in_tile.png`

![image](https://github.com/ImageMagick/ImageMagick/assets/71090115/5765a840-2ee4-449a-beb3-db76b0b08d9d)

**Comment by snibgo** [comment, marked as answer]
We can do a "saliency-blend" like this. "Saliency" means, roughly speaking, high-frequency detail. So each small area of the output will contain the high-frequency detail of either the truck or the background, depending on which contains the most detail. Windows syntax:

```
magick ^
  target_img.field512.png ^
  ( +clone obj_square.png ^
    -compose Over -composite ^
  ) ^
  obj_square.mask.png -negate ^
  -compose saliency_blend -composite ^
  sal_blend.png
```
Within the parentheses, we compose the truck over the background. Then we do the saliency blend within the mask, so the output within the mask contains detail of the background where the truck has no detail, or detail from the truck where that has strong detail.

sal_blend.png is:

![sal_blend](https://github.com/ImageMagick/ImageMagick/assets/33812368/de5d303a-2171-48d1-a5a6-726706bf6614)

Provided the mask is bigger than the truck, it doesn't make a difference. For example, this gives the same result:
```
%IMG7%magick ^
  target_img.field512.png ^
  ( +clone obj_square.png ^
    -compose Over -composite ^
  ) ^
  ( +clone -fill White -colorize 100 ) ^
  -compose saliency_blend -composite ^
  sal_blend2.png
```
sal_blend2.png is:

![sal_blend2](https://github.com/ImageMagick/ImageMagick/assets/33812368/0bef943c-cfb5-4ca1-a8b8-324eb9cfef6b)

If we shrink the mask so it doesn't include all of the outline of the truck, those parts won't appear in the output. However, this causes low-frequency detail from the truck to bleed outwards.
```
%IMG7%magick ^
  target_img.field512.png ^
  ( +clone obj_square.png ^
    -compose Over -composite ^
  ) ^
  ( obj_square.mask.png ^
    -morphology erode disk:3 ^
    -alpha off ^
    +write msk.png ^
  ) ^
  -compose saliency_blend -composite ^
  sal_blend3.png
```
sal_blend3.png is:

![sal_blend3](https://github.com/ImageMagick/ImageMagick/assets/33812368/1452df15-608c-424f-8f3e-9ca22259bea9)

Seamless blending is more like an alpha blending. The effect is obvious when we use the eroded mask. Again, low-frequency detail from the truck bleeds outwards, this time thoughout the entire image.
```
%IMG7%magick ^
  target_img.field512.png ^
  ( +clone obj_square.png ^
    -compose Over -composite ^
  ) ^
  ( obj_square.mask.png ^
    -morphology erode disk:3 ^
    -alpha off ^
    +write msk.png ^
  ) ^
  -compose seamless_blend -composite ^
  smlss_blend.png
```
smlss_blend.png is:

![smlss_blend](https://github.com/ImageMagick/ImageMagick/assets/33812368/1daeb1ac-752c-40b0-8afb-07939babf06f)

Does that help?

EDIT: I included "`+write msk.png`" as a debugging aid, so we can see the actual mask that is used. It has no effect on the output, so you can remove that.

**Comment by druzhynin-oleksii** [comment]
Thanks a lot. smlss_blend.png looks very close to what we need. 

How can I join https://www.imagemagick.org/discourse-server/viewforum.php?f=1 to contact you for some more specific private consultation?

**Comment by snibgo** [comment]
That forum is a read-only archive. (I don't generally encourage emails.)

If you haven't already seen it, you might find [Seamless photomontage](http://im.snibgo.com/seamlpm.htm) useful. I wrote it before IM developed the seamless-blend and saliency-blend operations. My code (Windows BAT) is much slower, but can use guided images, which allows for non-constant gradients.

**Reply by fmw42** [reply]
snibgo,  I think you meant earthlink, not earthling

**Reply by snibgo** [reply]
No, I didn't mean that.

**Reply by fmw42** [reply]
Sorry, never heard of earthling dot net.  Thought you meant earthlink dot net.  My mistake.

**Reply by snibgo** [reply]
You have used the address I showed, many times.

**Comment by druzhynin-oleksii** [comment]
Hi Alan.

Thanks a lot for sharing working command. Could you take a look how to get rid of a white artifact in another sample from our dataset:

** obj_in_tile.png **
![image](https://github.com/ImageMagick/ImageMagick/assets/71090115/448c64b4-d420-445b-badf-411808bf6c5d)


I got this after running

```shell
magick -verbose \
  target_img.field512.png \
  \( +clone obj_square.png \
    -compose Over -composite \
  \) \
  \( obj_square.mask.png \
    -morphology erode disk:3 \
    -alpha off \
  \) \
  -compose seamless_blend -composite \
  obj_in_tile.png
```

on the following images

**obj_square.png**
![image](https://github.com/ImageMagick/ImageMagick/assets/71090115/6e11d1a7-ee75-4b74-8e10-0ba551fd9052)

**obj_square.mask.png**
![image](https://github.com/ImageMagick/ImageMagick/assets/71090115/5b44e1f0-cb82-4c11-85d4-590dc449b729)

**target_img.field512.png**
![image](https://github.com/ImageMagick/ImageMagick/assets/71090115/9aa0ca0f-4a11-4b76-bb29-e2535ec8001f)


Regards,

Oleksii

**Comment by snibgo** [comment]
The white mark you highlight is "inside" the mask. That is, corresponding pixels in the mask are white.

Where the mask is black, corresponding pixels won't change (much). Where the mask is white, corresponding pixels will change, in an attempt to create a seamless blend between the two images.

Where did `obj_square.mask.png` come from? It can contain black pixels where you don't want changes. For example, we can ignore that file, and instead use the extracted alpha from `obj_square.png`. Windows syntax:
```
%IMG7%magick -verbose ^
  target_img.field512.png ^
  ( +clone ( obj_square.png +write mpr:INP ) ^
    -compose Over -composite ^
  ) ^
  ( mpr:INP -alpha extract ^
    -morphology erode disk:3 ^
    -alpha off ^
  ) ^
  -compose seamless_blend -composite ^
  obj_in_tile2.png
```
![obj_in_tile2](https://github.com/ImageMagick/ImageMagick/assets/33812368/4e2898cc-62d6-42ca-a631-93a6f14cb70a)

**Comment by druzhynin-oleksii** [comment]
Thanks a lot.

It seems like there is a corner case glitch.

![image](https://github.com/ImageMagick/ImageMagick/assets/71090115/6ace0f7f-641b-4438-82a2-195a85f02046)

this happens only at the following modified **obj_square.png**

![image](https://github.com/ImageMagick/ImageMagick/assets/71090115/05307765-af30-46d1-a5ff-7e0f3db13606)

This happens only if the dark pixels touch the boundary of the object.  
![image](https://github.com/ImageMagick/ImageMagick/assets/71090115/889b3d12-c28d-4cd8-b4e8-f24b972d9617)

**Comment by snibgo** [comment]
I suggest first adding a transparent border 1 pixel wide around obj_square:
```
magick -verbose ^
  target_img.field512.png ^
  ( +clone ( obj_square2.png -bordercolor None -border 1 +write mpr:INP ) ^
    -compose Over -composite ^
  ) ^
  ( mpr:INP -alpha extract ^
    -morphology erode disk:3 ^
    -alpha off ^
  ) ^
  -compose seamless_blend -composite ^
  obj_in_tile4.png
```

![obj_in_tile4](https://github.com/ImageMagick/ImageMagick/assets/33812368/e638a8f4-c5c4-41e4-a694-1b7d60b7855d)

**Comment by druzhynin-oleksii** [comment]
Hi Alan.

Thanks a lot for providing working solution.

Could you also help how to code
```
    # magick -verbose \
    #  target_img.field512.png \
    #  \( +clone \( obj_square.png +write mpr:INP \) \
    #    -compose Over -composite \
    #  \) \
    #  \( mpr:INP -alpha extract \
    #    -morphology erode disk:3 \
    #    -alpha off \
    #  \) \
    #  -compose seamless_blend -composite \
    #  obj_in_tile.png
```
in wand. I am trying 
```
    target_img = wand.image.Image.from_array(target_rgba, channel_map='RGBA')
    src_img = wand.image.Image.from_array(obj_square_rgba, channel_map='RGBA')
    target_img_clone = target_img.clone()

    # ( +clone ( obj_square.png +write mpr:INP ) -compose Over -composite )
    src_img_clone = src_img.clone()
    target_img_clone.composite(src_img, operator='over', arguments='')

    # ( mpr:INP -alpha extract -morphology erode disk:3 -alpha off )
    src_img_clone.alpha_channel = 'extract'
    src_img_clone.morphology(method='erode', kernel='disk:3')
    src_img_clone.alpha_channel = 'off'

    # -compose seamless_blend -composite
    target_img.composite(target_img_clone, operator='seamless_blend', arguments='')
```

The problem how to pass the mask (e.g. https://stackoverflow.com/questions/18938614/composition-masking-using-python-wand)

I've already added https://github.com/emcconville/wand/issues/626

**Comment by snibgo** [comment]
Sorry, I don't use Python, so can't help with that.

**Comment by druzhynin-oleksii** [comment]
Hi Alan.

Can we just do without `\( +clone obj_square.png -compose Over -composite \)`

```
    magick -verbose \
      larger_img.png \
      \( fit_smaller.png +write mpr:INP \) \
      \( mpr:INP -alpha extract \
       -morphology erode disk:3 \
        -alpha off \
      \) \
      -compose seamless_blend -composite \
      blended.png
```

**Comment by snibgo** [comment]
If it gives the result you want, of course you can do that.

**Comment by druzhynin-oleksii** [comment]
Yes,

```
 magick -verbose \
  target_img.field512.png \
  \( obj_square.png +write mpr:INP \) \
  \( mpr:INP -alpha extract \
     -morphology erode disk:3 \
     -alpha off \
  \) \
  -compose seamless_blend -composite \
  obj_in_tile.png
```
gives
![image](https://github.com/ImageMagick/ImageMagick/assets/71090115/d4260a8a-db1e-4a2f-bb60-093fd19e5c64)

I am just trying to understand the purpose of
```
  ( +clone obj_square2.png  ^
    -compose Over -composite ^
  ) ^
```

**Comment by druzhynin-oleksii** [comment]
is `mpr:INP` in-memory image with id `INP`?

**Comment by snibgo** [comment]
At any stage in a command, you can insert operations like "`+write x0.png`" to write the current image list. Do this as many times as you want, to help you understand the command. So you might have:
```
  ( +clone obj_square2.png  ^
    -compose Over -composite ^
    +write x0.png ^
  ) ^
```
... and examine x0.png and obj_square2.png, and compare them. In this case, it composites the square image over the field512 image. 

> is mpr:INP in-memory image with id INP?

Yes. I usually use capitals for the id, but we don't have to.

**Comment by druzhynin-oleksii** [comment]
about Python version: https://github.com/emcconville/wand/discussions/629

**Comment by druzhynin-oleksii** [comment]
Hi Alan.

Could you help to understand why 

**target_img.field512.png**
![image](https://github.com/ImageMagick/ImageMagick/assets/71090115/8e419750-0fd6-4e4b-88c0-14ec5f8a5757)

**obj_small.png**
![image](https://github.com/ImageMagick/ImageMagick/assets/71090115/95594e18-30ca-4a68-a7bf-e7af15131445)

```
magick -verbose \
  target_img.field512.png \
  \( obj_small.png +write mpr:INP \) \
  mpr:INP \
  -compose seamless_blend -composite \
  obj_in_tile.png
```
gives just a white artifact in the left top corner
![image](https://github.com/ImageMagick/ImageMagick/assets/71090115/aa7f84d2-32c6-40fb-b5b2-87b33990afca)

**Comment by druzhynin-oleksii** [comment]
```
magick -verbose \
  target_img.field512.png \
  \( obj_small.png +write mpr:INP \) \
  \( mpr:INP -alpha extract \
     -morphology erode disk:3 \
     -alpha off \
  \) \
  -compose seamless_blend -composite \
  obj_in_tile.png
```
left only couple pixels of the eroded object

![image](https://github.com/ImageMagick/ImageMagick/assets/71090115/01d4069f-5b09-4394-be76-94d0bb53c0e7)

![image](https://github.com/ImageMagick/ImageMagick/assets/71090115/0a2dafa1-2f55-4dbc-bc70-d0804dabcf40)

**Comment by snibgo** [comment]
> Could you help to understand why ... gives just a white artifact in the left top corner

The problem is that obj_small.png contains transparent pixels. Worse, they are transparent *black*. Hence there is a steep colour gradient between the (transparent) black and the other (opaque) pixels.

An improvement is to replace transparent *black* with transparent *average*, where *average* is the mean of the opaque pixels.
```
magick ^
  obj_small.png ^
  ( +clone -alpha extract +write mpr:ALP +delete ) ^
  ( +clone -scale "1x1^!" -scale "14x11^!" ) ^
  -compose DstOver -composite ^
  mpr:ALP ^
  -alpha off -compose CopyOpacity -composite ^
  bl.png
```
bl.png contains the same opaque pixels, and the same transparency, as obj_small.png, but the transparent colour is now approx sRGB(36%,40%,41%).

We can then use this:
```
magick ^
  target_img.field512.png ^
  ( bl.png +write mpr:INP ) ^
  mpr:INP ^
  -compose seamless_blend -composite ^
  x5.png
```

> left only couple pixels of the eroded object

The object is not eroded. 

You have eroded the mask, making it smaller. So the transparent black pixels in the object are further outside the boundary of the mask, so they don't matter so much.

With a smaller mask, fewer pixels from obj_small are used.

**Comment by druzhynin-oleksii** [comment]
thanks, looks nice

![image](https://github.com/ImageMagick/ImageMagick/assets/71090115/164a2969-9ab8-4bd4-a68d-070b30c6b872)

the right one is just copy-paste of obj_small.png 

but if I apply to **obj_square.png**
```
magick \
  obj_square.png \
  \( +clone -alpha extract +write mpr:ALP +delete \) \
  \( +clone -scale "1x1^!" -scale "14x11^!" \) \
  -compose DstOver -composite \
  mpr:ALP \
  -alpha off -compose CopyOpacity -composite \
  obj_square_bl.png


magick \
  target_img.field512.png \
  \( obj_square_bl.png +write mpr:INP \) \
  mpr:INP \
  -compose seamless_blend -composite \
  obj_in_tile.png
```
the result seems too violet
![image](https://github.com/ImageMagick/ImageMagick/assets/71090115/73170447-cf1a-44a9-bb2d-996300a5921a)

**Comment by druzhynin-oleksii** [comment]
Can we move the following just to separate image (need it to convert the code to Python):
```
  obj_small.png ^
  ( +clone -alpha extract +write mpr:ALP +delete ) ^
  ( +clone -scale "1x1^!" -scale "14x11^!" ) ^
  -compose DstOver -composite ^
```

**Comment by snibgo** [comment]
I cloned `obj_small.png`, scaled it to 1x1 (which makes one pixel that is the average of the image), then scaled that up to 14x11 because that is the size of `obj_small.png`.

`obj_square.png` is larger than 14x11. It is 455x177 pixels.

Caret ^ is the Windows escape character. You use bash, so don't use that. In bash, just do this:
```
-scale 1x1! -scale 455x177!
```

**Comment by druzhynin-oleksii** [comment]
Can we substitute?

```
magick \
  obj_small.png \
  \( +clone -alpha extract +write mpr:ALP +delete \) \
  \( +clone -scale "1x1^!" -scale "14x11^!" \) \
  -compose DstOver -composite \
  mpr:ALP \
  -alpha off -compose CopyAlpha -composite \
  obj_small_bl.png
```

with
```
magick obj_small.png -alpha extract obj_small_alpha.png 

magick \
  obj_small.png \
  \( +clone -scale "1x1^!" -scale "14x11^!" \) \
  -compose DstOver -composite \
  obj_small_bl1.png
  
magick \
  obj_small_bl1.png \
  obj_small_alpha.png \
  -alpha off -compose CopyAlpha -composite \
  obj_small_bl.png
```

**Comment by snibgo** [comment]
Yes, commands can be split, or combined.

**Reply by druzhynin-oleksii** [reply]
Why -`alpha off` is needed in `-alpha off -compose CopyAlpha -composite \`.

**Comment by snibgo** [comment]
`-composite` takes two images. We want `-compose CopyAlpha -composite` to copy the alpha channel of the first image from the grayscale of the second image. However, if the second image has an alpha channel, then that would be copied, instead of the grayscale. `-alpha off` ensures the second image has no alpha.

**Comment by druzhynin-oleksii** [comment]
but what `magick obj_small.png -alpha extract obj_small_alpha.png ` does? copies obj_small.png without alpha channel?

`  \( +clone -alpha extract +write mpr:ALP +delete \) \`

**Comment by snibgo** [comment]
`-alpha extract` copies the alpha channel to the active colour channels. See https://imagemagick.org/script/command-line-options.php#alpha

In your case, `-alpha off` is superfluous because the second image has no alpha channel. It is a very cheap operation: IM sees it has no alpha channel, so it does nothing.

**Comment by druzhynin-oleksii** [comment]
So, `\( +clone -alpha extract +write mpr:ALP +delete \) \` gives a gray-scale image with one channel create from alpha of obj_small.png?
